parsl/executors/globus_compute.py

```diff
@@ -0,0 +1,166 @@
+from __future__ import annotations
+
+import copy
+import uuid
+from concurrent.futures import Future
+from typing import Any, Callable, Dict, Optional, Union
+
+from parsl.errors import OptionalModuleMissing
+from parsl.executors.base import ParslExecutor
+from parsl.utils import RepresentationMixin
+
+try:
+    from globus_compute_sdk import Client, Executor
+    _globus_compute_enabled = True
+except ImportError:
+    _globus_compute_enabled = False
+
+UUID_LIKE_T = Union[uuid.UUID, str]
+
+
+class GlobusComputeExecutor(ParslExecutor, RepresentationMixin):
+    """ GlobusComputeExecutor enables remote execution on Globus Compute endpoints
+
+    GlobusComputeExecutor is a thin wrapper over globus_compute_sdk.Executor
+    Refer to `globus-compute user documentation <https://globus-compute.readthedocs.io/en/latest/executor.html>`_
+    and `reference documentation <https://globus-compute.readthedocs.io/en/latest/reference/executor.html>`_
+    for more details.
+
+    .. note::
+       As a remote execution system, Globus Compute relies on serialization to ship
+       tasks and results between the Parsl client side and the remote Globus Compute
+       Endpoint side. Serialization is unreliable across python versions, and
+       wrappers used by Parsl assume identical Parsl versions across on both sides.
+       We recommend using matching Python, Parsl and Globus Compute version on both
+       the client side and the endpoint side for stable behavior.
+
+    """
+
+    def __init__(
+            self,
+            endpoint_id: Optional[UUID_LIKE_T] = None,
+            task_group_id: Optional[UUID_LIKE_T] = None,
+            resource_specification: Optional[Dict[str, Any]] = None,
+            user_endpoint_config: Optional[Dict[str, Any]] = None,
+            label: str = "GlobusComputeExecutor",
+            batch_size: int = 128,
+            amqp_port: Optional[int] = None,
+            client: Optional[Client] = None,
+            **kwargs,
+    ):
+        """
+        Parameters
+        ----------
+
+        endpoint_id:
+            id of the endpoint to which to submit tasks
```

Comment:
what happens when this is `None`?

Reply:
I'll add an assertion to force this. There's a discrepancy between the executor and the `compute_sdk.Executor` in that an endpoint_id is required at init since we don't have a mechanism in Parsl to update the `endpoint_id` dynamically at runtime.

Reply:
I've added typeguard to `__init__` to cover this.